modules/localization/msf/local_pyramid_map/pyramid_map/pyramid_map_node.cc

```diff
@@ -0,0 +1,484 @@
+/******************************************************************************
+ * Copyright 2019 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+#include "modules/localization/msf/local_map/pyramid_map/pyramid_map_node.h"
+
+#include <vector>
+#include "modules/localization/msf/local_map/pyramid_map/pyramid_map_matrix.h"
+#include "modules/localization/msf/local_map/pyramid_map/pyramid_map_matrix_handler.h"
+#include "modules/localization/msf/local_map/pyramid_map/pyramid_map_node_config.h"
+
+namespace apollo {
+namespace localization {
+namespace msf {
+
+PyramidMapNode::PyramidMapNode() {}
+
+PyramidMapNode::~PyramidMapNode() {}
+
+void PyramidMapNode::Init(const BaseMapConfig* map_config) {
+  map_config_ = map_config;
+
+  map_node_config_ = new PyramidMapNodeConfig();
+  map_node_config_->map_version_ = map_config_->GetMapVersion();
+  if (map_node_config_->map_version_ == MapVersion::LOSSY_FULL_ALT_MAP ||
+      map_node_config_->map_version_ == MapVersion::LOSSLESS_MAP) {
+    map_node_config_->has_map_version_ = false;
+    map_node_config_->has_body_md5_ = false;
+  } else {
+    map_node_config_->has_map_version_ = true;
+    map_node_config_->has_body_md5_ = true;
+  }
+
+  is_reserved_ = false;
+  data_is_ready_ = false;
+  is_changed_ = false;
+
+  map_matrix_ = new PyramidMapMatrix();
+  map_matrix_handler_ =
+      PyramidMapMatrixHandlerSelector::AllocPyramidMapMatrixHandler(
+          map_node_config_->map_version_);
+  compression_strategy_ = new ZlibStrategy();
+
+  const PyramidMapConfig* pm_map_config =
+      dynamic_cast<const PyramidMapConfig*>(map_config_);
+  resolutions_mr_.resize(pm_map_config->resolution_num_, 0);
+  for (unsigned int i = 0; i < resolutions_mr_.size(); ++i) {
+    resolutions_mr_[i] = static_cast<float>(pm_map_config->map_resolutions_[0] *
+                         std::pow(pm_map_config->resolution_ratio_, i));
+  }
+
+  InitMapMatrix(map_config_);
+}
+
+void PyramidMapNode::Init(const BaseMapConfig* map_config,
+                          const MapNodeIndex& index, bool create_map_cells) {
+  map_config_ = map_config;
+
+  map_node_config_ = new PyramidMapNodeConfig();
```

Comment:
Use safe pointer. It applies to all "new" pointers.

Reply:
fixed